File: scala-parser/src/main/scala/Parsers.scala
package parsers

import scala.annotation.tailrec
import scala.util.matching.Regex
import scala.util.{Failure, Success, Try}

case class ParseResult[T](
    result: T,
    remainder: String
)

type Parser[T] = String => Try[ParseResult[T]]

def string(s: String): Parser[String] =
  input =>
    if input.startsWith(s) then
      Success(ParseResult(s, input.substring(s.length)))
    else Failure(IllegalArgumentException(s"expected $s"))

def regex(r: Regex): Parser[String] =
  input =>
    r.findPrefixMatchOf(input) match
      case Some(m) => Success(ParseResult(m.matched, input.substring(m.end)))
      case None    => Failure(IllegalArgumentException(s"expected $r"))

extension [T, R](p: Parser[T])
  def map(f: T => Try[R]): Parser[R] =
    input =>
      for
        ParseResult(intermediate, remainder) <- p(input)
        result <- f(intermediate)
      yield ParseResult(result, remainder)

def seq2[T1, T2](p1: Parser[T1], p2: Parser[T2]): Parser[(T1, T2)] =
  input =>
    for
      ParseResult(r1, remainder) <- p1(input)
      ParseResult(r2, remainder) <- p2(remainder)
    yield ParseResult((r1, r2), remainder)

def seq3[T1, T2, T3](
    p1: Parser[T1],
    p2: Parser[T2],
    p3: Parser[T3]
): Parser[(T1, T2, T3)] =
  input =>
    for
      ParseResult(r1, remainder) <- p1(input)
      ParseResult(r2, remainder) <- p2(remainder)
      ParseResult(r3, remainder) <- p3(remainder)
    yield ParseResult((r1, r2, r3), remainder)

class ExpectedOneOfException(val exceptions: List[Throwable]) extends Exception

def any[T](parsers: Parser[T]*): Parser[T] =
  input =>
    val results = parsers.map(_(input))
    val (successes, failures) = parsers
      .map(_(input))
      .partitionMap(_ match
        case Success(value)     => Left(value)
        case Failure(exception) => Right(exception)
      )
    (successes.headOption, failures) match
      case (Some(result), _) => Success(result)
      case (_, failures)     => Failure(ExpectedOneOfException(failures.toList))

def skip[T1, T2](p1: Parser[T1], p2: Parser[T2]): Parser[T2] =
  input =>
    for ParseResult((_, result), remainder) <- seq2(p1, p2)(input)
    yield ParseResult(result, remainder)

def bracketed[T1, T2, T3](
    p1: Parser[T1],
    p2: Parser[T2],
    p3: Parser[T3]
): Parser[T2] =
  input =>
    for ParseResult((_, result, _), remainder) <- seq3(p1, p2, p3)(input)
    yield ParseResult(result, remainder)

enum RangeOptions {
  case Unbounded
  case Bounded(value: Int)
}

case class Range(val lower: RangeOptions, val upper: RangeOptions) {
  def contains(x: Int): Boolean =
    (lower, upper) match
      case (RangeOptions.Unbounded, RangeOptions.Unbounded)      => true
      case (RangeOptions.Bounded(lower), RangeOptions.Unbounded) => x >= lower
      case (RangeOptions.Unbounded, RangeOptions.Bounded(upper)) => x <= upper
      case (RangeOptions.Bounded(lower), RangeOptions.Bounded(upper)) =>
        x >= lower && x <= upper
}

def repeat[T](p: Parser[T], r: Range): Parser[List[T]] =
  /*
	recursively match the next element and update the partial results as we go
	returns either a list of results that satisfy the range, or none
   */
  @tailrec
  def helper(
      input: String,
      allResultsSoFar: List[T],
      lastSuccessfulResults: Option[List[T]]
  ): Try[ParseResult[List[T]]] =
    // if adding one more result to total results would put us past the end of the range then we're done
    if r.contains(allResultsSoFar.length) && !r.contains(
        allResultsSoFar.length + 1
      )
    then
      lastSuccessfulResults match
        // we had some previous success
        case Some(x) => Success(ParseResult(x, input))
        // we never had a success so just try to be descriptive for what we were looking for
        case None =>
          Failure(
            IllegalArgumentException(
              s"didn't get enough matches for $p, valid range $r"
            )
          )
    else
      // see if we can actually make another match
      p(input) match
        case Success(ParseResult(newResult, remainder)) =>
          // we have a new result
          val newAllResults = newResult :: allResultsSoFar
          // if that new list works then keep it, otherwise whatever we last had as the successful list
          // e.g. we might be skipping some because our range has an increment other than 1
          val newLastSuccessfulResults =
            if r.contains(newAllResults.length) then Some(newAllResults)
            else lastSuccessfulResults
          // recurse
          helper(remainder, newAllResults, newLastSuccessfulResults)
        case Failure(e) =>
          // failed to match again, just return whatever we have
          lastSuccessfulResults match
            // we had some previous success
            case Some(x) => Success(ParseResult(x, input))
            // we never had a success so just try to be descriptive for what we were looking for
            case None =>
              Failure(
                java.lang.IllegalArgumentException(
                  s"not enough results, was looking for $r, but didn't get enough matches"
                )
              )

  // if the range allows for an empty result as a success then we can start with that
  val initialSuccessResults = if r.contains(0) then Some(List()) else None
  input => helper(input, List(), initialSuccessResults)

// TODO optional

File: swift-parser/Sources/parsers.swift
public struct ParseResult<T> {
	public let result: T
	public let remainder: Substring

	public init(result: T, remainder: Substring) {
		self.result = result
		self.remainder = remainder
	}
}

extension ParseResult: Equatable where T: Equatable {}

public struct ParseError: Error, Equatable {
	public init() {}
}

public protocol Parser<T> {
	associatedtype T

	func apply(input: Substring) -> Result<ParseResult<T>, ParseError>
}

struct MapParser<SourceType, ReturnType>: Parser {
	typealias T = ReturnType

	let p: any Parser<SourceType>
	let f: (SourceType) -> Result<ReturnType, ParseError>

	func apply(input: Substring) -> Result<ParseResult<ReturnType>, ParseError> {
		p(input: input)
			.flatMap { result in
				f(result.result).map { newResult in
					ParseResult(result: newResult, remainder: result.remainder)
				}
			}
	}
}

extension Parser {
	public func apply(input: String) -> Result<ParseResult<T>, ParseError> {
		self.apply(input: Substring(input))
	}

	public func callAsFunction(input: Substring) -> Result<ParseResult<T>, ParseError> {
		self.apply(input: input)
	}

	public func map<R>(f: @escaping (T) -> Result<R, ParseError>) -> any Parser<R> {
		MapParser(p: self, f: f)
	}
}

struct StringParser: Parser {
	typealias T = String

	let s: String

	func apply(input: Substring) -> Result<ParseResult<String>, ParseError> {
		if input.starts(with: s) {
			.success(
				ParseResult(
					result: s,
					remainder: input.dropFirst(s.count)
				))
		} else {
			.failure(ParseError())
		}
	}
}

public func string(_ s: String) -> any Parser<String> {
	StringParser(s: s)
}

struct RegexParser: Parser {
	typealias T = Substring

	let r: Regex<Substring>

	func apply(input: Substring) -> Result<ParseResult<Substring>, ParseError> {
		switch try? r.prefixMatch(in: input) {
		case .none:
			.failure(ParseError())
		case .some(let result):
			.success(
				ParseResult(result: result.output, remainder: input[result.output.endIndex...]))
		}
	}
}

public func regex(_ r: Regex<Substring>) -> any Parser<Substring> {
	RegexParser(r: r)
}

struct Seq2Parser<T1, T2>: Parser {
	typealias T = (T1, T2)

	let p1: any Parser<T1>
	let p2: any Parser<T2>

	func apply(input: Substring) -> Result<ParseResult<(T1, T2)>, ParseError> {
		do {
			let r1 = try p1(input: input).get()
			let r2 = try p2(input: r1.remainder).get()
			return .success(
				ParseResult(
					result: (r1.result, r2.result),
					remainder: r2.remainder
				))
		} catch {
			return .failure(error)
		}
	}
}

public func seq2<T1, T2>(
	_ p1: any Parser<T1>,
	_ p2: any Parser<T2>
) -> any Parser<(T1, T2)> {
	Seq2Parser(p1: p1, p2: p2)
}

struct Seq3Parser<T1, T2, T3>: Parser {
	typealias T = (T1, T2, T3)

	let p1: any Parser<T1>
	let p2: any Parser<T2>
	let p3: any Parser<T3>

	func apply(input: Substring) -> Result<ParseResult<(T1, T2, T3)>, ParseError> {
		do {
			let r1 = try p1(input: input).get()
			let r2 = try p2(input: r1.remainder).get()
			let r3 = try p3(input: r2.remainder).get()
			return .success(
				ParseResult(
					result: (r1.result, r2.result, r3.result),
					remainder: r3.remainder
				))
		} catch {
			return .failure(error)
		}
	}
}

public func seq3<T1, T2, T3>(
	_ p1: any Parser<T1>,
	_ p2: any Parser<T2>,
	_ p3: any Parser<T3>
) -> any Parser<(T1, T2, T3)> {
	Seq3Parser(p1: p1, p2: p2, p3: p3)
}

struct AnyParser<T>: Parser {
	let parsers: [any Parser<T>]

	func apply(input: Substring) -> Result<ParseResult<T>, ParseError> {
		for p in parsers {
			if case .success(let result) = p(input: input) {
				return .success(result)
			}
		}
		return .failure(ParseError())
	}
}

public func any<T>(_ parsers: any Parser<T>...) -> any Parser<T> {
	AnyParser(parsers: parsers)
}

public func skip<T1, T2>(_ p1: any Parser<T1>, _ p2: any Parser<T2>) -> any Parser<T2> {
	seq2(p1, p2).map { result in .success(result.1) }
}

public func bracketed<T1, T2, T3>(_ p1: any Parser<T1>, _ p2: any Parser<T2>, _ p3: any Parser<T3>)
	-> any Parser<T2>
{
	seq3(p1, p2, p3).map { result in .success(result.1) }
}

// TODO repeat
// TODO many0
// TODO many1
// TODO optional
